domain/secret/state/state_test.go

```diff
@@ -3643,3 +3643,86 @@ func (s *stateSuite) TestGetObsoleteUserSecretRevisionsReadyToPrune(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(result, jc.SameContents, []string{uri.ID + "/1"})
 }
+
+func (s *stateSuite) TestChangeSecretBackend(c *gc.C) {
+	st := newSecretState(c, s.TxnRunnerFactory())
+	ctx := context.Background()
+
+	s.setupUnits(c, "mysql")
+	uriCharm1 := coresecrets.NewURI()
+	uriCharm2 := coresecrets.NewURI()
+	uriUser1 := coresecrets.NewURI()
+	uriUser2 := coresecrets.NewURI()
+
+	dataInput := coresecrets.SecretData{"foo": "bar", "hello": "world"}
+	valueRefInput := &coresecrets.ValueRef{
+		BackendID:  "backend-id",
+		RevisionID: "revision-id",
+	}
+
+	err := st.CreateCharmApplicationSecret(ctx, 1, uriCharm1, "mysql", domainsecret.UpsertSecretParams{Data: dataInput})
+	c.Assert(err, jc.ErrorIsNil)
+	data, valueRef, err := st.GetSecretValue(ctx, uriCharm1, 1)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(data, jc.DeepEquals, dataInput)
+	c.Assert(valueRef, gc.IsNil)
+
+	err = st.CreateCharmApplicationSecret(ctx, 1, uriCharm2, "mysql", domainsecret.UpsertSecretParams{ValueRef: valueRefInput})
+	c.Assert(err, jc.ErrorIsNil)
+	data, valueRef, err = st.GetSecretValue(ctx, uriCharm2, 1)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(data, gc.IsNil)
+	c.Assert(valueRef, gc.DeepEquals, valueRefInput)
+
+	err = st.CreateUserSecret(ctx, 1, uriUser1, domainsecret.UpsertSecretParams{Data: dataInput})
+	c.Assert(err, jc.ErrorIsNil)
+	data, valueRef, err = st.GetSecretValue(ctx, uriUser1, 1)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(data, jc.DeepEquals, dataInput)
+	c.Assert(valueRef, gc.IsNil)
+
+	err = st.CreateUserSecret(ctx, 1, uriUser2, domainsecret.UpsertSecretParams{ValueRef: valueRefInput})
+	c.Assert(err, jc.ErrorIsNil)
+	data, valueRef, err = st.GetSecretValue(ctx, uriUser2, 1)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(data, gc.IsNil)
+	c.Assert(valueRef, gc.DeepEquals, valueRefInput)
+
+	// Change the backend failed.
```

Comment:
Would prefer separate tests for things like this, eg error checking as opposed to happy path